Brevity #274 is over

ixmanuel opened the pull request:
@yegor256 CountOf ready. If you prefer I can rename it to LengthOf

- CountOf(Text, String) counts the elements.
- CountOf(Iterable, Iterator) counts the elements.
- CountOf(Input) the size in bytes.
- CountOf(Bytes) the size in bytes.

alex-semenyuk commented on src/main/java/org/cactoos/CountOf.java:
```diff
@@ -0,0 +1,125 @@
+ * @version $Id$
+ * @since 0.12
+ */
+public final class CountOf implements Scalar<Number> {
```
@ixmanuel why Number instead of Integer?

svendiedrichsen replied:
.. or instead of Long?

ixmanuel replied:
Hi!, yes @alex-semenyuk, for counting elements integer is ok and for bytes as you said @svendiedrichsen, long is applicable. CountOf returns integer for the fist one and long for the latter one. That's why it returns Number. It is show it in some of its tests.

alex-semenyuk commented on src/main/java/org/cactoos/CountOf.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * The underlying scalar number.
+     */
+    private final UncheckedScalar<?> scalar;
```
@ixmanuel maybe need explicit parameterized that do not type casting 

> (Number) this.scalar.value();

ixmanuel replied:
@alex-semenyuk yes, it is an alternative but, I tried to make it clear and "ergonomic" for the end-user-developer. Alternatively, we could split counting for bytes and elements into: CountOf(Text, String, Iterable, Iterator) and SizeOf(Input, Bytes).

ixmanuel replied:
@alex-semenyuk The ```Scalar<?>``` is because it can accept any type of ```Iterable<?>``` and ```Iterator<?>```

ixmanuel commented:
### Final ## Report

_Note: The variations in params are not mentioned, just the main objects involved._
```
### Text
- TextOf
  String, Reader, StringBuilder, StringBuffer, char..., Throwable, 
  Input and Bytes.

- JoinedText
- RepeatedText
- ReversedText
- SubText

- CountOf
  String, Text

### Input
- InputOf
  File, Path, URI, URL, Scalar<URL>, Reader, StringBuilder, 
  StringBuffer, char..., String, Text, Throwable, byte[], 
  Bytes and InputStream.

- ResourceOf
  String

- CountOf
  Input

### Bytes
- BytesOf: 
  Input, Reader, StringBuilder, StringBuffer, String, Text, 
  char..., Throwable, and byte...

- CountOf
  Bytes

### List
- PropertiesOf
  String, Text, Input, Map.Entry and Map     

- ArrayOf
  Map, X...

- ListOf
- MapOf
- CharactersOf

- JoindeOf
- MappedOf
- ReducedOf
- FilteredOf  
- CycledOf
- ReversedOf 
- SkippedOf
- SortedOf
- StickyOf
- SubOf

- CountOf
  Iterable, Iterator

- StickyList
- StickyMap

- RepeatedItem
- EndlessItem
- ItemAt

### Fun
- CallableOf
- FuncOf
- MatcherOf
- ProcOf
- RunnableOf

### Pending
FileAsOutput    |    OutputOf (1)
OutputStreamAsOutput | OutputOf (1)
PathAsOutput    |    OutputOf (1)

SumOfInts       |    SumOf (2)
SumOfReals      |    SumOf (2)
```

ixmanuel commented:
@yegor256, @fabriciofx, @svendiedrichsen, @alex-semenyuk:

# Final Report

_Note: The variations in params are not mentioned, just the main types involved. The untouched are also not mentioned._

## text
- TextOf
  _String, Reader, StringBuilder, StringBuffer, char..., Throwable, 
  Input and Bytes._
- JoinedText, RepeatedText, ReversedText and SubText.
- CountOf
  _String, Text._

## io
- InputOf
  _File, Path, URI, URL, Scalar<URL>, Reader, StringBuilder, 
  StringBuffer, char..., String, Text, Throwable, byte[], 
  Bytes and InputStream._
- ResourceOf
- CountOf
  _Input._
- OutputTo
  _File, Path, OutputStream._

## bytes
- BytesOf: 
  _Input, Reader, StringBuilder, StringBuffer, String, Text, 
  char..., Throwable, and byte..._
- CountOf
  _Bytes._

## list
- PropertiesOf
  _String, Text, Input, Map.Entry and Map._     
- ArrayOf
  _Map, X..._
- ListOf, MapOf, CharactersOf.
- JoindeOf, MappedOf, ReducedOf, FilteredOf, CycledOf, ReversedOf,
  SkippedOf, SortedOf, StickyOf, SubOf.
- CountOf
  _Iterable, Iterator._
- StickyList, StickyMap.
- RepeatedItem, EndlessItem, ItemAt.
- SumOfInts and SumOfReals.*  
_* I can try with declared interfaces within the object SumOf in order to avoid same erasure_

## func
- CallableOf, FuncOf, MatcherOf, ProcOf, RunnableOf.

## docs
- Readme

yegor256 commented:
@ixmanuel I think `LengthOf` is better than `CountOf`. Pls, rename it.

ixmanuel commented:
@yegor256
### Report
- I think that iterators should not be visible for the API too and we could keep the package name: **list** or might be renamed it to **collection**.
- I think we could rename package **func** by **scalar**: It has funcs, procs, equals, booleans, scalars and I changed TextAs{Scalar}  to {Scalar}Of. Now them, have the same spirit as TextOf.
- I changed and renamed StringAsUrl and UrlAsString to io package with the names: EncodedUrl and DecodedUrl.
- LengthOf is back again.
- The name for "mutators" has remained as in the initial proposal and it has been modified any other object not considered too. I refer to "mutators" to the objects that modify the source (in a new object) but return the same type. It could be a guideline for upcoming objects.

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/12539-318983570)

rultor commented:
> @rultor try to merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/12539-318983570) (took me 10min)
